core/src/main/scala/org/apache/spark/storage/BlockInfoManager.scala

```diff
@@ -0,0 +1,356 @@
+    // What about block transfer / getRemote()?
+    Option(TaskContext.get()).map(_.taskAttemptId()).getOrElse(-1024L)
+  }
+
+  /**
+   * Lock a block for reading and return its metadata.
+   *
+   * A single task can lock a block multiple times for reading, in which case each lock will need
+   * to be released separately.
+   *
+   * @param blockId the block to lock.
+   * @param blocking if true (default), this call will block until the lock is acquired. If false,
+   *                 this call will return immediately if the lock acquisition fails.
+   * @return None if the block did not exist or was removed (in which case no lock is held), or
+   *         Some(BlockInfo) (in which case the block is locked for reading).
+   */
+  def lockForReading(
+      blockId: BlockId,
+      blocking: Boolean = true): Option[BlockInfo] = synchronized {
+    logTrace(s"Task $currentTaskAttemptId trying to acquire read lock for $blockId")
+    infos.get(blockId).map { info =>
+      while (info.writerTask != -1) {
+        if (info.removed) return None
+        if (blocking) wait() else return None
+      }
+      if (info.removed) return None
+      info.readerCount += 1
+      readLocksByTask(currentTaskAttemptId).add(blockId)
+      logTrace(s"Task $currentTaskAttemptId acquired read lock for $blockId")
+      info
+    }
+  }
+
+  /**
+   * Lock a block for writing and return its metadata.
+   *
+   * If this is called by a task which already holds the block's exclusive write lock, then this
+   * will return success but will not further increment any lock counts (so both write-lock
+   * acquisitions will be freed by the same [[unlock()]] or [[downgradeLock()]] call.
+   *
+   * @param blockId the block to lock.
+   * @param blocking if true (default), this call will block until the lock is acquired. If false,
+   *                 this call will return immediately if the lock acquisition fails.
+   * @return None if the block did not exist or was removed (in which case no lock is held), or
+   *         Some(BlockInfo) (in which case the block is locked for writing).
+   */
+  def lockForWriting(
+      blockId: BlockId,
+      blocking: Boolean = true): Option[BlockInfo] = synchronized {
+    logTrace(s"Task $currentTaskAttemptId trying to acquire write lock for $blockId")
+    infos.get(blockId).map { info =>
+      if (info.writerTask != currentTaskAttemptId) {
+        while (info.writerTask != -1 || info.readerCount != 0) {
+          if (info.removed) return None
+          if (blocking) wait() else return None
+        }
+        if (info.removed) return None
+      }
+      info.writerTask = currentTaskAttemptId
+      writeLocksByTask.addBinding(currentTaskAttemptId, blockId)
+      logTrace(s"Task $currentTaskAttemptId acquired write lock for $blockId")
+      info
+    }
+  }
+
+  /**
+   * Get a block's metadata without acquiring any locks.
+   */
+  def get(blockId: BlockId): Option[BlockInfo] = synchronized {
```

Comment:
This is used in the implementation of `BlockManager.getStatus`, which is why it's public. I can make this `private[storage]` and add a comment explaining this.